File: src/bmsx_cpp/machine/devices/vdp/blitter_source.cpp
#include "machine/devices/vdp/blitter_source.h"

#include "machine/devices/device_status.h"
#include "machine/devices/vdp/contracts.h"
#include "machine/devices/vdp/vram.h"

namespace bmsx {

VdpBlitterSourcePort::VdpBlitterSourcePort(DeviceStatusLatch& fault, VdpVramUnit& vram)
	: m_fault(fault)
	, m_vram(vram) {}

bool VdpBlitterSourcePort::resolveSurfaceIdForSlot(u32 slot, u32& surfaceId, u32 faultCode) {
	if (slot == VDP_SLOT_SYSTEM) {
		surfaceId = VDP_RD_SURFACE_SYSTEM;
		return true;
	}
	if (slot == VDP_SLOT_PRIMARY) {
		surfaceId = VDP_RD_SURFACE_PRIMARY;
		return true;
	}
	if (slot == VDP_SLOT_SECONDARY) {
		surfaceId = VDP_RD_SURFACE_SECONDARY;
		return true;
	}
	m_fault.raise(faultCode, slot);
	return false;
}

VdpSurfaceUploadSlot* VdpBlitterSourcePort::resolveSlotSurface(u32 slot, u32 faultCode) {
	u32 surfaceId = 0u;
	if (!resolveSurfaceIdForSlot(slot, surfaceId, faultCode)) {
		return nullptr;
	}
	VdpSurfaceUploadSlot* surface = m_vram.findSurface(surfaceId);
	if (surface == nullptr) {
		m_fault.raise(faultCode, surfaceId);
		return nullptr;
	}
	return surface;
}

bool VdpBlitterSourcePort::resolveWordsInto(u32 slot, u32 u, u32 v, u32 w, u32 h, VdpBlitterSource& target, u32 faultCode) {
	u32 surfaceId = 0u;
	if (!resolveSurfaceIdForSlot(slot, surfaceId, faultCode)) {
		return false;
	}
	target.surfaceId = surfaceId;
	target.srcX = u;
	target.srcY = v;
	target.width = w;
	target.height = h;
	return true;
}

bool VdpBlitterSourcePort::validateSurface(const VdpBlitterSource& source, u32 faultCode, u32 zeroSizeFaultCode) {
	if (source.width == 0u || source.height == 0u) {
		m_fault.raise(zeroSizeFaultCode, source.width | (source.height << 16u));
		return false;
	}
	const VdpSurfaceUploadSlot* surface = m_vram.findSurface(source.surfaceId);
	if (surface == nullptr) {
		m_fault.raise(faultCode, source.surfaceId);
		return false;
	}
	const uint64_t sourceRight = static_cast<uint64_t>(source.srcX) + static_cast<uint64_t>(source.width);
	const uint64_t sourceBottom = static_cast<uint64_t>(source.srcY) + static_cast<uint64_t>(source.height);
	if (sourceRight > surface->surfaceWidth || sourceBottom > surface->surfaceHeight) {
		m_fault.raise(faultCode, source.srcX | (source.srcY << 16u));
		return false;
	}
	return true;
}

} // namespace bmsx

File: src/bmsx/machine/devices/vdp/blitter_source.ts
import type { DeviceStatusLatch } from '../device_status';
import {
	VDP_RD_SURFACE_PRIMARY,
	VDP_RD_SURFACE_SECONDARY,
	VDP_RD_SURFACE_SYSTEM,
	VDP_SLOT_PRIMARY,
	VDP_SLOT_SECONDARY,
	VDP_SLOT_SYSTEM,
} from './contracts';
import type { VdpBlitterSource } from './blitter';
import type { VdpSurfaceUploadSlot } from './device_output';
import type { VdpVramUnit } from './vram';

export class VdpBlitterSourcePort {
	public constructor(
		private readonly fault: DeviceStatusLatch,
		private readonly vram: VdpVramUnit,
	) {}

	private readonly surfaceScratch = new Uint32Array(1);

	public resolveSurfaceIdForSlot(slot: number, out: Uint32Array, faultCode: number): boolean {
		if (slot === VDP_SLOT_SYSTEM) {
			out[0] = VDP_RD_SURFACE_SYSTEM;
			return true;
		}
		if (slot === VDP_SLOT_PRIMARY) {
			out[0] = VDP_RD_SURFACE_PRIMARY;
			return true;
		}
		if (slot === VDP_SLOT_SECONDARY) {
			out[0] = VDP_RD_SURFACE_SECONDARY;
			return true;
		}
		this.fault.raise(faultCode, slot);
		return false;
	}

	public resolveSlotSurface(slot: number, faultCode: number): VdpSurfaceUploadSlot | null {
		if (!this.resolveSurfaceIdForSlot(slot, this.surfaceScratch, faultCode)) {
			return null;
		}
		const surfaceId = this.surfaceScratch[0]!;
		const surface = this.vram.findSurface(surfaceId);
		if (surface === null) {
			this.fault.raise(faultCode, surfaceId);
			return null;
		}
		return surface;
	}

	public resolveWordsInto(slot: number, u: number, v: number, w: number, h: number, target: VdpBlitterSource, faultCode: number): boolean {
		if (!this.resolveSurfaceIdForSlot(slot, this.surfaceScratch, faultCode)) {
			return false;
		}
		target.surfaceId = this.surfaceScratch[0]!;
		target.srcX = u;
		target.srcY = v;
		target.width = w;
		target.height = h;
		return true;
	}

	public validateSurface(source: VdpBlitterSource, faultCode: number, zeroSizeFaultCode: number): boolean {
		if (source.width === 0 || source.height === 0) {
			this.fault.raise(zeroSizeFaultCode, (source.width | (source.height << 16)) >>> 0);
			return false;
		}
		const surface = this.vram.findSurface(source.surfaceId);
		if (surface === null) {
			this.fault.raise(faultCode, source.surfaceId);
			return false;
		}
		if (source.srcX + source.width > surface.surfaceWidth || source.srcY + source.height > surface.surfaceHeight) {
			this.fault.raise(faultCode, (source.srcX | (source.srcY << 16)) >>> 0);
			return false;
		}
		return true;
	}
}
